### ISSUE TITLE
[dev.icinga.com #12669] Usage of regex in assignments rules

### ISSUE THREAD

**Issue by icinga-migration** [Icinga/icingaweb2-module-director#435]
This issue has been migrated from Redmine: https://dev.icinga.com/issues/12669

**Created by Foxeronie on 2016-09-07 14:02:26 +00:00**

Assignee: _(none)_
Status: _New_ 
Target Version: _(none)_
Last Update: _2016-09-07 14:02:26 +00:00 (in Redmine)_

---

Hi!

It would be great if regex support like "^it-puppet0\[1-6\]" for assign rules were possible via director. :)

Best regards,
Patrick

**Comment by choonming** [comment]
@Thomas-Gelf would this be implemented anytime soon? using the match conditions are not that feasible in our use case now

**Comment by ClemensBW** [comment]
+1 :)

**Comment by vlk-charles** [comment]
+1

**Comment by menewol** [comment]
this is the shit!!!!!!! we need it!!!! :O

**Comment by dkatona** [comment]
Missing regex support in apply rules limits my use of icinga director or it leads to hacks like assigning a custom variable to each service/host just to be able to filter by that variable in apply rules. This issue is opened for almost a year and a half, is there any timeline when it can be fixed?

**Comment by cstegm** [comment]
We need this!

**Comment by doctore74** [comment]
Standard in other products. We need this asap.

**Comment by menewol** [comment]
Since we're shotstorming at it again...we've worked around this by creating a customVar that has to be set for every host (very dirty and un-automated).

So: this is still the Shit!!! We need it!

**Comment by DRhyme** [comment]
We need these freakin regex assignment rules right now.

**Comment by mtdeguzis** [comment]
I am running into this now. A splat (*) auto migrates to a match rule, but ?? does not

```
object HostGroup "prod-mgt-nodes" {
    display_name = "prod-mgt-nodes"
    assign where match("site1*prlx.domain.com", host.name)
}
object HostGroup "prod-cluster" {
    display_name = "prod-cluster"
    assign where host.name == "site2??prlx.domain.com"
}
```

**Comment by Thomas-Gelf** [comment]
Blocker for this is that we use `Icinga\Data\Filter` to persist filter rules, and they have no related syntax. Consistent behavior involving equal regex matches on PostgreSQL, MySQL and in the GUI might also become tricky, but should be doable. Introducing `?` for our "normal" matching rules, eventually translating them to regex for Icinga would be easier, but then we should eventually introduce an escaping mechanism.

For imported objects (like from PuppetDB) it's usually easy to work around this issue: use an Import modifier to add dedicated flags based on regular expressions.

**Comment by nomoresecrets** [comment]
Is there any news on this issue?

**Comment by Thomas-Gelf** [comment]
Nope, there is still no generic regular expression support in the Icinga Web 2 filter syntax. And I strongly recommend anyone to set Host properties (custom variables) wherever you feel the desire to assign something based on apply rules.

Site, datacenter, location, operating system - none of the examples I have been presented qualifies for a plain regex. You'll end up with the same regular expression duplicated in multiple places, a co-worker  might tweak only one of them and forget about the other one... I really love regular expressions, but I personally do not miss them in Apply Rule expressions at all. Do yourself a favor and give your hosts meaningful properties.

For manually created hosts, provide related fields in your templates. For imported hosts import those properties. Or use the Regex property modifier at import time to fill additional columns with related values based on a regular expression.

If you really want to try to move this feature request forward, you might consider opening a feature request in the Icinga Web 2 repository. But even with the willingness to sponsor this, please note that this will be a complex feature, requiring a lot of work:

* expressions need to be transformed to at least MySQL and PostgreSQL
* Oracle, MSSQL and DB2 would optionally also ask for related changes. We usually do not write to them, but we support them in extra modules and as import sources
* they need to behave the same way in the database, in Icinga 2 and via PHP / in memory
* this COULD require us to define a supported subset of functionality, but we wouldn't have any chance to validate whether a given filter expression respects such restriction
* what happens when such a filter hits an LDAP connection, as this could eventually happen?
* charset support and case sensitiveness must behave the same way on all platforms

Sure, all of this is possible. But it is not an easy task and could lead to a huge technical debt and many potential issues when done in a rush.

We could eliminate some of above concerns (there might be others) by either splitting up the filter string parser into modules. The filter itself is modular, the parser not.Then we could opt for a different subset of features in different places of our code. Of course we could also fork the filter code into the Director module, so that we could apply changes with less side-effects. But even then, most of the mentioned issues still need to be addressed.

**Comment by luciov66** [comment]
is there any news on this issue? we need a way to configure assign rules able to pick up hosts with a hostname complying with a regex of this type 
<string1> <any N characters> <string2>
for instance, if N=2:
host12name and host34name should be included by the rule;
host67pippo and host55ciao should not be included by the rule;
there is no way at the moment to configure this.
thanks.

**Comment by djkerya** [comment]
+1
